File: advertising_sdk/src/main/java/com/adadapted/android/sdk/AdAdapted.java
package com.adadapted.android.sdk;

import android.content.Context;
import android.util.Log;

import com.adadapted.android.sdk.config.Config;
import com.adadapted.android.sdk.core.addit.Content;
import com.adadapted.android.sdk.core.addit.PayloadClient;
import com.adadapted.android.sdk.core.event.AppEventClient;
import com.adadapted.android.sdk.core.session.Session;
import com.adadapted.android.sdk.core.session.SessionClient;
import com.adadapted.android.sdk.ext.Wireup;
import com.adadapted.android.sdk.ui.messaging.AaSdkAdditContentListener;
import com.adadapted.android.sdk.ui.messaging.AaSdkEventListener;
import com.adadapted.android.sdk.ui.messaging.AaSdkSessionListener;
import com.adadapted.android.sdk.ui.messaging.AdditContentPublisher;
import com.adadapted.android.sdk.ui.messaging.SdkEventPublisher;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AdAdapted {
    private static final String LOGTAG = AdAdapted.class.getName();

    public static class Env {
        public static final boolean PROD = true;
        public static final boolean DEV = false;
    }

    private static AdAdapted sInstance;

    private boolean hasStarted;
    private String appId;
    private boolean isProd;
    private Map<String, String> params;
    private AaSdkSessionListener sessionListener;
    private AaSdkEventListener eventListener;
    private AaSdkAdditContentListener contentListener;

    private AdAdapted() {
        hasStarted = false;
        params = new HashMap<>();
    }

    private synchronized static AdAdapted getsInstance() {
        if(sInstance == null) {
            sInstance = new AdAdapted();
        }

        return sInstance;
    }

    public static AdAdapted init() {
        return getsInstance();
    }

    public AdAdapted withAppId(final String appId) {
        if(appId == null) {
            Log.e(LOGTAG, "The Application Id cannot be Null.");
            this.appId = "";
        }
        else {
            this.appId = appId;
        }

        return this;
    }

    public AdAdapted inEnv(final boolean environment) {
        isProd = environment;

        return this;
    }

    public AdAdapted setSdkSessionListener(final AaSdkSessionListener listener) {
        sessionListener = listener;

        return this;
    }

    public AdAdapted setSdkEventListener(final AaSdkEventListener listener) {
        eventListener = listener;

        return this;
    }

    public AdAdapted setSdkAdditContentListener(final AaSdkAdditContentListener listener) {
        contentListener = listener;

        return this;
    }

    public AdAdapted withParams(final Map<String, String> params) {
        this.params = params;

        return this;
    }

    public void start(final Context context) {
        if(context == null) {
            Log.e(LOGTAG, "The Context cannot be NULL");
            return;
        }

        if(appId == null) {
            Log.e(LOGTAG, "The Application Id cannot be NULL");
            return;
        }

        if(hasStarted) {
            Log.w(LOGTAG, "AdAdapted Android Advertising SDK has already been started");
            return;
        }

        hasStarted = true;

        Wireup.run(context, isProd);

        SdkEventPublisher.getInstance().setListener(eventListener);
        AdditContentPublisher.getInstance().addListener(contentListener);

        PayloadClient.pickupPayloads(new PayloadClient.Callback() {
            @Override
            public void onPayloadAvailable(final List<Content> content) {
                if(content.size() > 0) {
                    AdditContentPublisher.getInstance().publishContent(content.get(0));
                }
            }
        });

        final SessionClient.Listener startListener = new SessionClient.Listener() {
            @Override
            public void onSessionAvailable(final Session session) {
                if(sessionListener != null) {
                    sessionListener.onHasAdsToServe(session.hasActiveCampaigns());
                }
            }

            @Override
            public void onAdsAvailable(Session session) {
                if(sessionListener != null) {
                    sessionListener.onHasAdsToServe(session.hasActiveCampaigns());
                }
            }

            @Override
            public void onSessionInitFailed() {
                if(sessionListener != null) {
                    sessionListener.onHasAdsToServe(false);
                }
            }
        };

        SessionClient.start(
            context.getApplicationContext(),
            appId,
            isProd,
            params,
            startListener);

        AppEventClient.trackSdkEvent("app_opened");

        Log.i(LOGTAG, String.format("AdAdapted Android Advertising SDK v%s initialized.", Config.SDK_VERSION));
    }

    public static synchronized void restart(final Context context) {
        restart(context, new HashMap<String, String>());
    }

    public static synchronized void restart(final Context context,
                                            final Map<String, String> params) {
        if(context == null) {
            Log.e(LOGTAG, "The Context cannot be NULL");
            return;
        }

        if(getsInstance().appId == null) {
            Log.e(LOGTAG, "The Application Id cannot be Null.");
            return;
        }

        AppEventClient.trackSdkEvent("sdk_restarted");

        SessionClient.restart(
            context.getApplicationContext(),
            getsInstance().appId,
            getsInstance().isProd,
            params
        );

        Log.i(LOGTAG, String.format("AdAdapted Android Advertising SDK v%s reinitialized.", Config.SDK_VERSION));
    }

    public static synchronized void hasAdsToServe() {
        if(getsInstance().sessionListener != null) {
            final Session session = SessionClient.getCurrentSession();

            if(session != null) {
                getsInstance().sessionListener.onHasAdsToServe(session.hasActiveCampaigns());
            }
        }
    }

    public static synchronized void registerEvent(final String eventName) {
        registerEvent(eventName, new HashMap<String, String>());
    }

    public static synchronized void registerEvent(final String eventName,
                                                  final Map<String, String> eventParams) {
        AppEventClient.trackAppEvent(eventName, eventParams);
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/AdAdapted.kt
package com.adadapted.android.sdk

import android.content.Context
import androidx.lifecycle.ProcessLifecycleOwner
import com.adadapted.android.sdk.constants.Config
import com.adadapted.android.sdk.core.ad.AdClient
import com.adadapted.android.sdk.core.atl.AddItContentPublisher
import com.adadapted.android.sdk.core.concurrency.Transporter
import com.adadapted.android.sdk.core.device.DeviceInfo
import com.adadapted.android.sdk.core.device.DeviceInfoClient
import com.adadapted.android.sdk.core.device.DeviceInfoExtractor
import com.adadapted.android.sdk.core.event.EventBroadcaster
import com.adadapted.android.sdk.core.event.EventClient
import com.adadapted.android.sdk.core.interfaces.AaSdkAdditContentListener
import com.adadapted.android.sdk.core.interfaces.AaSdkEventListener
import com.adadapted.android.sdk.core.interfaces.DeviceCallback
import com.adadapted.android.sdk.core.keyword.InterceptClient
import com.adadapted.android.sdk.core.keyword.KeywordInterceptMatcher
import com.adadapted.android.sdk.core.log.AALogger
import com.adadapted.android.sdk.core.network.HttpAdAdapter
import com.adadapted.android.sdk.core.network.HttpConnector
import com.adadapted.android.sdk.core.network.HttpEventAdapter
import com.adadapted.android.sdk.core.network.HttpInterceptAdapter
import com.adadapted.android.sdk.core.network.HttpPayloadAdapter
import com.adadapted.android.sdk.core.payload.PayloadClient
import com.adadapted.android.sdk.core.session.SessionClient

object AdAdapted {
    enum class Env { PROD, DEV }
    private var hasStarted = false
    private var apiKey: String = ""
    private var isProd = false
    private var customIdentifier: String = ""
    private var isKeywordInterceptEnabled = false
    private var isPayloadEnabled = false
    private var params: Map<String, String> = HashMap()
    private lateinit var eventListener: AaSdkEventListener
    private lateinit var contentListener: AaSdkAdditContentListener

    fun withAppId(key: String): AdAdapted {
        this.apiKey = key
        return this
    }

    fun inEnv(env: Env): AdAdapted {
        isProd = env == Env.PROD
        return this
    }

    fun enableKeywordIntercept(value: Boolean): AdAdapted {
        isKeywordInterceptEnabled = value
        return this
    }

    fun enablePayloads(value: Boolean): AdAdapted {
        isPayloadEnabled = value
        return this
    }

    fun setSdkEventListener(listener: AaSdkEventListener): AdAdapted {
        eventListener = listener
        return this
    }

    fun setSdkAdditContentListener(listener: AaSdkAdditContentListener): AdAdapted {
        contentListener = listener
        return this
    }

    fun setOptionalParams(params:HashMap<String, String>): AdAdapted {
        this.params = params
        return this
    }

    fun enableDebugLogging(): AdAdapted {
        AALogger.enableDebugLogging()
        return this
    }

    fun setCustomIdentifier(identifier: String): AdAdapted {
        customIdentifier = identifier
        return this
    }

    fun disableAdTracking(context: Context): AdAdapted {
        setAdTracking(context, true)
        return this
    }

    fun enableAdTracking(context: Context): AdAdapted {
        setAdTracking(context, false)
        return this
    }

    fun start(context: Context) {
        if (apiKey.isEmpty()) {
            AALogger.logError("The AdAdapted Api Key is missing or NULL")
        }
        if (hasStarted) {
            return
        }
        hasStarted = true
        setupClients(context)
        eventListener.let { EventBroadcaster.setListener(it) }
        contentListener.let { AddItContentPublisher.addListener(it) }
        AALogger.logInfo("AdAdapted Android SDK ${Config.LIBRARY_VERSION} initialized.")
    }

    private fun setAdTracking(context: Context, value: Boolean) {
        val sharedPref =
            context.getSharedPreferences(Config.AASDK_PREFS_KEY, Context.MODE_PRIVATE) ?: return
        with(sharedPref.edit()) {
            putBoolean(Config.AASDK_PREFS_TRACKING_DISABLED_KEY, value)
            apply()
        }
    }

    private fun setupClients(context: Context) {
        Config.init(isProd)

        val deviceInfoExtractor = DeviceInfoExtractor(context)
        DeviceInfoClient.createInstance(
            apiKey,
            isProd,
            params,
            customIdentifier,
            deviceInfoExtractor,
            Transporter()
        )

        DeviceInfoClient.getDeviceInfo(object : DeviceCallback {
            override fun onDeviceInfoCollected(deviceInfo: DeviceInfo) {
                setupDependentClients()
            }
        })

        ProcessLifecycleOwner.get().lifecycle.addObserver(SessionClient)
    }

    private fun setupDependentClients() {
        AdClient.createInstance(
            HttpAdAdapter(
                "https://dev.adadapted.dev/api/ad-service/v100-alpha/ad/retrieve", //TODO TEMP REMOVE
                //Config.getRetrieveAdsUrl(),
                HttpConnector
            ), Transporter()
        )
        EventClient.createInstance(
            HttpEventAdapter(
                Config.getAdEventsUrl(),
                Config.getSdkEventsUrl(),
                Config.getSdkErrorsUrl(),
                HttpConnector
            ), Transporter()
        )
        InterceptClient.createInstance(
            HttpInterceptAdapter(
                "https://dev.adadapted.dev/api/ad-service/v100-alpha/intercept/retrieve", //TODO TEMP REMOVE
                //Config.getRetrieveInterceptsUrl(),
                Config.getInterceptEventsUrl(),
                HttpConnector
            ), Transporter(),
            isKeywordInterceptEnabled
        )
        PayloadClient.createInstance(
            HttpPayloadAdapter(
                Config.getPickupPayloadsUrl(),
                Config.getTrackingPayloadUrl(),
                HttpConnector
            ), EventClient, Transporter()
        )

        if (isKeywordInterceptEnabled) {
            KeywordInterceptMatcher.match("INIT") //init the matcher
        }

        if (isPayloadEnabled) {
            PayloadClient.pickupPayloads {
                if (it.isNotEmpty()) {
                    for (content in it) {
                        AddItContentPublisher.publishAddItContent(content)
                    }
                }
            }
        }
    }
}